### PR TITLE
Nucleo l552ze

### PR DESCRIPTION
**Pull request by kenbell**
Fixes https://github.com/tinygo-org/tinygo/issues/1531

Support for the ST nucleo l552ze.  UART and LED functionality has been verified.

Note `machine_stm32_l5_uart.go` implementation is identical to STM F7 implementation.  I've done a separate file, but would be possible to combine / merge.

### CODE REVIEW

**Comment by ardnew** [comment]
To match convention with other existing STM32 peripheral file names, as well as matching the `stm32l5` build tag, I recommend renaming the source file:

`machine_stm32_l5_uart.go` -> `machine_stm32l5_uart.go`

Unfortunately it looks like the F7 source files were already merged with this same scheme (#1526), but it too should probably be renamed `machine_stm32f7_uart.go` to match its respective build tag as well

**Comment by ardnew** [comment]
This UART implementation appears to be identical (excluding the `// +build` constraints) to the `machine_stm32_f7_uart.go` that was just merged. 

And these two are both nearly identical to the STM32 family-agnostic implementation in `machine_stm32_uart.go` ([which you very nicely explained here](https://github.com/tinygo-org/tinygo/pull/1526#issuecomment-743782343)):

```diff
--- a/machine_stm32_uart.go
+++ b/machine_stm32_f7_uart.go
@@ -1,4 +1,4 @@
-// +build stm32,!stm32f7
+// +build stm32f7

 package machine

@@ -42,7 +42,7 @@ func (uart UART) Configure(config UARTConfig) {
 // handleInterrupt should be called from the appropriate interrupt handler for
 // this UART instance.
 func (uart *UART) handleInterrupt(interrupt.Interrupt) {
-       uart.Receive(byte((uart.Bus.DR.Get() & 0xFF)))
+       uart.Receive(byte((uart.Bus.RDR.Get() & 0xFF)))
 }

 // SetBaudRate sets the communication speed for the UART. Defer to chip-specific
@@ -54,9 +54,9 @@ func (uart UART) SetBaudRate(br uint32) {

 // WriteByte writes a byte of data to the UART.
 func (uart UART) WriteByte(c byte) error {
-       uart.Bus.DR.Set(uint32(c))
+       uart.Bus.TDR.Set(uint32(c))

-       for !uart.Bus.SR.HasBits(stm32.USART_SR_TXE) {
+       for !uart.Bus.ISR.HasBits(stm32.USART_ISR_TXE) {
        }
        return nil
 }
```

Regardless if these are combined with the family-agnostic implementation, e.g., abstracting the register names as you mentioned, the L5 and F7 sources can and _should_ be combined as-is, without abstraction.

I'm having trouble imagining a naming convention to use for these family-combined sources. The following table shows the currently-supported TinyGo targets (including this PR) and which files implement their UART functionality:

|Target|STM32 Family|ARM Core|`machine_stm32_uart.go`|`machine_stm32_f7_uart.go`|
|:------:|:------:|:----:|:--------------------------:|:------------------------------:|
|`bluepill`|STM32F1|Cortex-M3|✓||
|`nucleof103rb`|STM32F1|Cortex-M3|✓||
|`feather_stm32f405`|STM32F4|Cortex-M4F|✓||
|`stm32f4disco`|STM32F4|Cortex-M4F|✓||
|`nucleof722ze`|STM32F7|Cortex-M7F||✓|
|`nucleol552ze`|STM32L5|Cortex-M33||✓|

So both source file implementations support multiple **STM32 families** _and_ multiple **ARM cores**, meaning neither one of them can simply append a single, comprehensive build tag to their file name.

Therefore, considering all of these targets are using identical UART driver logic, I would recommend using your original suggestion to use abstracted registers, which can be defined in the *existing* `board_`/`machine_` source files of the respective targets -- probably as private/un-exported members in each of their `type UART struct` definitions. And then we can remove `machine_stm32_f7_uart.go`, and just have **all** of the STM32 targets use the one sole implementation in `machine_stm32_uart.go`.

**Comment by kenbell** [comment]
Sounds good - i'll take a pass at unifying the implementations.

**Comment by kenbell** [comment]
I've pushed a change aimed at harmonizing the UART logic.  I'm not sure in some of the choices I made, specifically:

I centralized the UART struct, so there's one definition for all STM32 variants - this means the f103xx now has an AltFuncSelector where it didn't before.  I think this is benign, but as a matter of taste may not be the right thing?

I've made all the UART member functions take UART* - after I centralized the definition of UART, I found some of the initialization code was changing a copy of the UART struct (which lead to some fun debugging :) )

File: targets/cortex-m33.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+	"inherits": ["cortex-m"],
+    "llvm-target": "armv7m-none-eabi",
```
Comment (deadprogram) [inline targets/cortex-m33.json:3]: Please adjust formatting.

File: src/machine/machine_stm32l5.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		stm32.RCC.APB1ENR1.SetBits(stm32.RCC_APB1ENR1_TIM6EN)
+	case unsafe.Pointer(stm32.TIM5): // TIM5 clock enable
+		stm32.RCC.APB1ENR1.SetBits(stm32.RCC_APB1ENR1_TIM5EN)
+	case unsafe.Pointer(stm32.TIM4): // TIM4 clock enable
+		stm32.RCC.APB1ENR1.SetBits(stm32.RCC_APB1ENR1_TIM4EN)
+	case unsafe.Pointer(stm32.TIM3): // TIM3 clock enable
+		stm32.RCC.APB1ENR1.SetBits(stm32.RCC_APB1ENR1_TIM3EN)
+	case unsafe.Pointer(stm32.TIM2): // TIM2 clock enable
+		stm32.RCC.APB1ENR1.SetBits(stm32.RCC_APB1ENR1_TIM2EN)
+
```
Comment (deadprogram) [inline src/machine/machine_stm32l5.go]: Is there a reason this line is blank?

**Comment by deadprogram** [comment]
This PR is looking pretty great. I made a few small comments about formatting. Consolidating as much code as possible for the various stm32 chips is a great objective, and I fully support this effort! :star:

File: targets/cortex-m33.json (hunk 1 above)
Comment (kenbell) [reply targets/cortex-m33.json:3]: File reformatted in change faa3524

File: src/machine/machine_stm32l5.go (hunk 2 above)
Comment (kenbell) [reply src/machine/machine_stm32l5.go]: File reformatted in change faa3524

**Comment by kenbell** [comment]
To adapt this change to the switch to stm32-rs SVDs needs a patch to that project to add GPIO enum values, see: https://github.com/stm32-rs/stm32-rs/pull/479

**Comment by kenbell** [comment]
I've raised a PR against stm32-svd to update with the GPIO enum values, see: https://github.com/tinygo-org/stm32-svd/pull/2

**Comment by deadprogram** [comment]
@kenbell the PR against the stm32-rs repo was merged. This PR should be ready to be able to be worked on again.

**Comment by kenbell** [comment]
@deadprogram - I think this one is good for review and/or merge.  I've verified on hardware with the latest stm32-svd.

**Comment by deadprogram** [comment]
Sorry @kenbell just one last merge conflict to be resolved, if you please.

**Comment by deadprogram** [comment]
I rebased and squashed some commits, and have merged into `dev`. Thanks for working on this @kenbell now closing as merged.

**Pull request closed without merging**